### PR TITLE
[Issue #2696] add New Relic to Next app

### LINKED ISSUE #2696: Generate NewRelic beacons from the frontend
### Summary

Implement in our client code
https://docs.newrelic.com/docs/browser/new-relic-browser/browser-pro-features/browser-data-distributed-tracing/

### Acceptance criteria

- [ ] Frontend has the NewRelic library
- [ ] Frontend is sending browser data requests using NewRelic library

### PR DESCRIPTION
**Pull request by doug-s-nava**
## Summary
Fixes #2696

## Changes proposed

The goal is to begin sending data to New Relic from the frontend Next application. Things that have been done:

* a New Relic entity called "Simpler Grants Next DEV" created to track Node and client / browser data
* code added as per direction from New Relic documentation and example Next app in order to add necessary fronted scripts for tracking timing data see:
    * https://www.npmjs.com/package/newrelic
    * https://github.com/newrelic/newrelic-node-examples/tree/58f760e828c45d90391bda3f66764d4420ba4990/nextjs-app-router
    * https://docs.newrelic.com/docs/apm/agents/nodejs-agent/installation-configuration/install-nodejs-agent-docker/ 
* custom typing added to plaster over deficiencies in supplied New Relic TS definitions
* support for necessary environment variables added in infra code
* tested locally as described below

### Test steps

1. start next on this branch with `NEW_RELIC_APP_NAME="<see 1pass>" NEW_RELIC_LICENSE_KEY="<see 1pass>" npm run dev:nr`
2. visit some pages and perform searches
1. _VERIFY_: data comes through into [New Relic](https://one.newrelic.com/nr1-core/page-views/home/NTI0OTgwOXxCUk9XU0VSfEFQUExJQ0FUSU9OfDExMjAzNTAwMjc?account=5249809&duration=1800000&filters=%28domain%20IN%20%28%27BROWSER%27%2C%20%27BROWSER%27%29%20AND%20type%20IN%20%28%27APPLICATION%27%2C%20%27MICRO_FRONTEND%27%29%29&state=7a8689e5-2312-28e4-34bc-98ae1f022400), in both APM and browser dashboards

### Screenshots after deploying to DEV
Screenshot 2024-11-22 at 10 04 24 AM
Screenshot 2024-11-22 at 10 03 50 AM

### Next Steps

* open an issue and possibly a PR with DefinitelyTyped to fix typing issues https://github.com/DefinitelyTyped/DefinitelyTyped
* deploy to DEV to confirm functionality in deployed environment
* confirm expected behavior in PROD (data comes in) and Staging (data does not come in) after deployment
* documentation

### CODE REVIEW

File: infra/frontend/app-config/env-config/environment-variables.tf [hunk 1]
```diff
@@ -43,6 +45,14 @@ locals {
...
       manage_method     = "manual"
       secret_store_name = "/${var.app_name}/${var.environment}/api-auth-token"
+    },
+    NEW_RELIC_APP_NAME = {
+      manage_method     = "manual"
+      secret_store_name = "/${var.app_name}/${var.environment}/new-relic-app-name"
+    },
+    NEW_RELIC_LICENSE_KEY = {
+      manage_method     = "manual"
+      secret_store_name = "new-relic-license-key"
```
[the remark this answers is not in the document]
Comment (doug-s-nava) [reply infra/frontend/app-config/env-config/environment-variables.tf:55]: Looks like these both work, which I'm gonna assume means that you don't need the slash, but happy to add it for uniformity if you want:
* `aws ssm get-parameter --name=new-relic-license-key`
* `aws ssm get-parameter --name=/new-relic-license-key`

File: infra/frontend/app-config/dev.tf [hunk 2]
```diff
@@ -15,4 +15,8 @@ module "dev_config" {
   instance_scaling_min_capacity   = 2
   # instance_scaling_max_capacity is 5x the instance_scaling_min_capacity
   instance_scaling_max_capacity = 10
+
+  service_override_extra_environment_variables = {
+    NEW_RELIC_ENABLED = "true"
```
Comment (coilysiren) [inline infra/frontend/app-config/dev.tf:20]: Is it intentional that you're using `false` (bool) versus `"true"` string?

File: infra/frontend/app-config/env-config/environment-variables.tf [hunk 3]
```diff
@@ -7,6 +7,8 @@ locals {
     # WORKER_THREADS_COUNT    = 4
     # LOG_LEVEL               = "info"
     # DB_CONNECTION_POOL_SIZE = 5
+    NEW_RELIC_ENABLED = false
```
Comment (coilysiren) [inline infra/frontend/app-config/env-config/environment-variables.tf:10]: Docs link please! Probably to here:

https://docs.newrelic.com/docs/apm/agents/nodejs-agent/installation-configuration/nodejs-agent-configuration/#agent-enabled

File: infra/frontend/app-config/prod.tf [hunk 4]
```diff
@@ -19,4 +19,9 @@ module "prod_config" {
 
   instance_cpu    = 1024
   instance_memory = 2048
+
+  service_override_extra_environment_variables = {
+    NEW_RELIC_ENABLED = "true"
+    NODE_OPTIONS      = "-r @newrelic/next"
```
Comment (coilysiren) [inline infra/frontend/app-config/prod.tf:25]: RE `NODE_OPTIONS`, can we just always input `-r @newrelic/next` and rely on the `NEW_RELIC_ENABLED` flag instead? It's gonna be unclear to people looking at the infra layer what they should be doing with `NODE_OPTIONS`

**Review by coilysiren — APPROVED** [review]
Obvs only reviewing "my" half of the PR, but I have optional feedback

File: infra/frontend/app-config/dev.tf (hunk 2 above)
Comment (doug-s-nava) [reply infra/frontend/app-config/dev.tf:20]: Oops! That mismatch is definitely a mistake. I wanted to follow the type definition here, so updating the `false` to null. https://github.com/HHS/simpler-grants-gov/blob/e9eed30cfc7a09c7851b0b85b4a4853aba774cd5/infra/frontend/app-config/env-config/variables.tf#L57

File: infra/frontend/app-config/prod.tf (hunk 4 above)
Comment (doug-s-nava) [reply infra/frontend/app-config/prod.tf:25]: yeah, I'm fine with that. In the end we'll have this enabled everywhere, and pre-loading the New Relic library, even if we aren't using it, shouldn't pose any problems.

File: infra/frontend/app-config/staging.tf [hunk 5]
```diff
@@ -20,4 +20,8 @@ module "staging_config" {
   instance_desired_instance_count = 4
   instance_scaling_min_capacity   = 4
   instance_scaling_max_capacity   = 20
+
+  service_override_extra_environment_variables = {
+    NEW_RELIC_ENABLED = "false"
+  }
```
Comment (coilysiren) [inline infra/frontend/app-config/staging.tf:26]: The current configuration sets the default to true, and then only staging to false. I don't think that's intentional?

Comment (coilysiren) [reply infra/frontend/app-config/staging.tf:26]: oh 🤦🏼‍♀️ idk if you were even ready for a re-review yet

Comment (coilysiren) [reply infra/frontend/app-config/staging.tf:26]: Kay!

**Review by coilysiren — APPROVED** [review]
🐑 it

**Review by acouch — APPROVED** [review]
👍🏿

**Pull request merged**